internal/globalconfig/globalconfig.go

```diff
@@ -106,15 +106,12 @@ func SetStatsCarrier(sc *internal.StatsCarrier) {
 // PushStat pushes the stat onto the StatsCarrier's stats channel, via the Add method
 func PushStat(stat internal.Stat) {
 	cfg.mu.RLock()
-	defer cfg.mu.RUnlock()
-	if !StatsCarrier() {
+	sc := cfg.statsCarrier
+	cfg.mu.RUnlock()
```

Comment:
Just to make sure I'm understanding - why are we unlocking here and not after the `if sc == nil` check? I understand that with `sc := cfg.statsCarrier`, we are assigning the globalconfig's statscarrier to a new local variable, `sc` and `sc` will not be changed by some other goroutine, but what if between lines 109 and 112, another goroutine does set a statscarrier, and `cfg. statsCarrier` is no longer nil even if `sc` is?
- I guess we still avoid a data race, but the stat will be dropped unnecessarily. Correct?
- But wouldn't moving the unlock to after the `if sc == nil` check help ensure better synchronization?

Reply:
Good questions! If another goroutine sets `cfg.statsCarrier`, it has to do it with a write lock. And it can't actually take the write lock until there are no other read or write locks held. So, even if we moved the unlock later in the function, this function would still observe whatever the value of `cfg.statsCarrier` was at the time, which could be nil.